src/compiler/interface.jl

```diff
@@ -139,7 +140,21 @@ end
 
 Base.show(io::IO, ps::Grads) = print(io, "Grads(...)")
 
-@forward Grads.grads Base.getindex, Base.haskey
+@forward Grads.grads  Base.haskey, Base.setindex!
+@forward Grads.params  Base.length
+
+# Dictionary interface.
+# Don't use the IdDict directly since it may contain some spurious pairs. 
+Base.keys(gs::Grads) = gs.params
+Base.values(gs::Grads) =  (gs.grads[p] for p in gs.params)
+Base.pairs(gs::Grads) =  (p => gs.grads[p] for p in gs.params)
```

Comment:
`Base.pairs` returns a `AbstractDict` subtype.
I think Base is wrong for doing that. Because you then can't call `map` on it.
But it is surprising that we diverge from this.

Reply:
turns out we don't have to define `pairs` explicitly, Base defines it generically from keys and values